File: hosts/android/feature/products/impl/src/main/java/io/paritytech/polkadotapp/feature_products_impl/domain/pocket/PublishedPocketCards.kt
package io.paritytech.polkadotapp.feature_products_impl.domain.pocket

import io.paritytech.polkadotapp.common.utils.flatMap
import io.paritytech.polkadotapp.feature_products_api.domain.pocket.PocketCardId
import io.paritytech.polkadotapp.feature_products_api.model.PocketCardDefinition
import io.paritytech.polkadotapp.feature_products_api.model.Product
import io.paritytech.polkadotapp.feature_products_api.model.ProductId
import io.paritytech.polkadotapp.feature_products_impl.domain.usecase.ResolveProductUseCase
import javax.inject.Inject

sealed class PocketPublishError(message: String) : Throwable(message) {
    data object NoPocket : PocketPublishError("product publishes no Pocket cards")

    data object UnknownCard : PocketPublishError("product publishes no such card")

    // Variants are singletons, so a captured trace would point at classloading, not the failure.
    override fun fillInStackTrace(): Throwable = this
}

data class PublishedPocketCard(
    val product: Product,
    val definition: PocketCardDefinition,
)

/** Looks a card up in the worker manifest of the product that claims to back it. */
class PublishedPocketCards @Inject constructor(
    private val resolveProductUseCase: ResolveProductUseCase,
) {
    suspend fun find(productId: ProductId, cardId: PocketCardId): Result<PublishedPocketCard> =
        resolveProductUseCase.resolve(productId).flatMap { resolved ->
            val worker = resolved.executables.worker
            if (worker == null || !worker.includesPocket) return@flatMap Result.failure(PocketPublishError.NoPocket)

            worker.pocketCards.firstOrNull { it.id == cardId }
                ?.let { Result.success(PublishedPocketCard(resolved.product, it)) }
                ?: Result.failure(PocketPublishError.UnknownCard)
        }
}

File: hosts/ios/polkadot-app/Modules/Products/Pocket/PublishedPocketCards.swift
import Foundation
import Products

/// Looks a card up in the worker manifest of the product that claims to back it.
///
/// Nothing here trusts the caller's spelling of the product: the resolver settles
/// on one canonical id, and that is the id the card is keyed by.
struct PublishedPocketCards: PublishedPocketCardsResolving {
    private let products: any ProductResolving
    private let debugCards: @Sendable (ProductId) -> [PocketCardDefinition]

    init(
        products: any ProductResolving,
        debugCards: @escaping @Sendable (ProductId) -> [PocketCardDefinition] = { _ in [] }
    ) {
        self.products = products
        self.debugCards = debugCards
    }

    func find(productId: ProductId, cardId: PocketCardId) async throws -> PublishedPocketCard {
        let resolved: ResolvedProduct
        do {
            resolved = try await products.resolve(productId)
        } catch {
            // A product that could not be read has not answered yet. A card
            // typed in by hand needs no chain presence and still stands;
            // anything else would settle the question against a read that
            // simply did not land.
            guard let byHand = byHand(cardId, of: productId, resolved: nil) else { throw error }

            return byHand
        }

        if let worker = resolved.executables.worker, worker.includesPocket {
            guard let definition = worker.pocketCards.first(where: { $0.id == cardId }) else {
                throw PocketPublishError.unknownCard
            }

            return PublishedPocketCard(
                productId: resolved.id,
                productName: resolved.displayName,
                workerContentId: worker.identifier,
                definition: definition
            )
        }

        // Only reached when the product publishes no Pocket worker: a published
        // one always wins, so a card supplied by hand can never shadow what a
        // product actually ships.
        guard let byHand = byHand(cardId, of: productId, resolved: resolved) else {
            throw PocketPublishError.noPocket
        }

        return byHand
    }

    private func byHand(
        _ cardId: PocketCardId,
        of productId: ProductId,
        resolved: ResolvedProduct?
    ) -> PublishedPocketCard? {
        guard let definition = debugCards(productId).first(where: { $0.id == cardId }) else { return nil }

        return PublishedPocketCard(
            productId: resolved?.id ?? productId,
            productName: resolved?.displayName ?? productId,
            workerContentId: resolved?.id ?? productId,
            definition: definition
        )
    }
}

extension PublishedPocketCards {
    /// The resolver the app uses: published manifests, with cards typed into the
    /// debug menu filling in for products that publish no worker.
    static func makeDefault(products: any ProductResolving) -> PublishedPocketCards {
        #if DEBUG
            let debug = DebugPocketCards()

            return PublishedPocketCards(products: products, debugCards: { debug.cards(for: $0) })
        #else
            return PublishedPocketCards(products: products)
        #endif
    }
}
